Description: make ptmx a bind mount rather than symlink
 udev on some systems checks the device type of /dev/ptmx, and replaces it if
 not as expected.  The symlink created by libvirt-lxc therefore gets replaced.
 By creating it as a bind mount, the device type is correct and udev leaves it
 alone.
Forwarded: yes
Author: Serge Hallyn <serge.hallyn@canonical.com>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/lxc/+bug/832123
---
 src/lxc/lxc_container.c |   20 ++++++++++----------
 1 files changed, 10 insertions(+), 10 deletions(-)

Index: libvirt-0.9.2/src/lxc/lxc_container.c
===================================================================
--- libvirt-0.9.2.orig/src/lxc/lxc_container.c	2011-09-01 09:16:01.000000000 -0500
+++ libvirt-0.9.2/src/lxc/lxc_container.c	2011-09-01 09:24:41.591278227 -0500
@@ -486,18 +486,18 @@
         }
     }
 
+    dev_t dev = makedev(LXC_DEV_MAJ_TTY, LXC_DEV_MIN_PTMX);
+    if (mknod("/dev/ptmx", S_IFCHR, dev) < 0 ||
+        chmod("/dev/ptmx", 0666)) {
+        virReportSystemError(errno, "%s",
+                             _("Failed to make device /dev/ptmx"));
+        return -1;
+    }
+
     if (access("/dev/pts/ptmx", W_OK) == 0) {
-        if (symlink("/dev/pts/ptmx", "/dev/ptmx") < 0) {
-            virReportSystemError(errno, "%s",
-                                 _("Failed to create symlink /dev/ptmx to /dev/pts/ptmx"));
-            return -1;
-        }
-    } else {
-        dev_t dev = makedev(LXC_DEV_MAJ_TTY, LXC_DEV_MIN_PTMX);
-        if (mknod("/dev/ptmx", S_IFCHR, dev) < 0 ||
-            chmod("/dev/ptmx", 0666)) {
+        if (mount("/dev/pts/ptmx", "/dev/ptmx", "ptmx", MS_BIND, NULL) < 0) {
             virReportSystemError(errno, "%s",
-                                 _("Failed to make device /dev/ptmx"));
+                                 _("Failed to bind-mount /dev/ptmx to /dev/pts/ptmx"));
             return -1;
         }
     }
